src/System/reachability.rs

```diff
@@ -0,0 +1,166 @@
+) -> Result<(), Box<dyn std::error::Error>> {
+    if !system.get_all_locations().contains(start_state.get_location())
+        {return Err("The transition system does not contain the start location".into())}
+    if !system.get_all_locations().contains(end_state.get_location())
+        {return Err("The transition system does not contain the end location".into())}
+
+    Ok(())
+}
+
+fn fails_early(
+    start_state: &State,
+    end_state: &State,
+    system: &dyn TransitionSystem
+) -> bool{
+    if let Some(invariants) = end_state.get_location().get_invariants() {
+        if !&end_state.zone_ref().has_intersection(invariants) {
+            return true;
+        }
+    }
+    false
+}
+///# Find path
+///
+/// Returns a path from a start state to an end state in a transition system.
+///
+/// If it is reachable, it returns a path.
+///
+/// If it is not reachable, it returns None.
+///
+/// The start state can be omitted with None to use the start state of the transition system.
+///
+///## Checking if a state can reach another:
+///
+/// let is_reachable: bool = match find_path(Some(start_state), end_state, transition_system) {
+///     Some(path) => true,
+///     None => false
+/// };
+///
+///## Omitting start state:
+///
+/// let is_reachable: bool = match find_path(None, end_state, transition_system) {
+///     Some(path) => true,
+///     None => false
+/// };
+///
+// This is the main function for the reachability query.
+pub fn find_path(
+    begin_state: Option<State>,
+    end_state: State,
+    system: &dyn TransitionSystem,
+) -> Result<Option<Path>, String> {
+    let start_state: State;
+    if let Some(s) = begin_state {
+        start_state = s;
+    } else if let Some(s) = system.get_initial_state() {
+        start_state = s;
+    } else {
+        panic!("No state to start with");
+    }
+
+    if let Err(err) = validate_input(&start_state, &end_state, system) {
+        return Err(err.to_string());
+    }
+
+    if fails_early(&start_state, &end_state, system){
+        return Ok(None);
+    }
+
+    search_algorithm(&start_state, &end_state, system)
+}
+
+pub fn search_algorithm(
+    start_state: &State,
+    end_state: &State,
+    system: &dyn TransitionSystem,
+) -> Result<Option<Path>, String> {
+
+    // hashmap linking every location to all its current zones
+    let mut visited_states:HashMap<LocationID, Vec<OwnedFederation>> = HashMap::new();
+
+    // List of states that are to be visited
+    let mut frontier_states: Vec<State> = Vec::new();
+
+    frontier_states.push(start_state.clone());
+    loop {
+        let next_state = frontier_states.pop();
+        // All has been explored if no next state exist
+        if next_state.is_none() {
+            break;
+        }
+        let next_state = next_state.unwrap();
+        if reached_end_state(&next_state, &end_state) {
+            return Ok(Some(Path{}))/* TODO: Return the actual path */
+        }
+        for action in system.get_actions(){
```

Comment:
Make instead a local variable outside the loop of actions. since every time  get_actions() is called it combine the input actions and output actions. Also the actions would not be changed, so therefore a local variable would be more efficient

Reply:
Fixed, it is now saved as a local variable outside the loop